### PR TITLE
Azure storage filesystem support

### PR DESCRIPTION
**Pull request by spacedragon**
```
export AZURE_STORAGE_KEY=<storage key>
tritonserver --model-repository=as://<storage account>.blob.core.windows.net/<container>/<path>
```

### CODE REVIEW

File: src/core/filesystem.cc [hunk 1]
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
...
+  }
+}
+
+Status ASFileSystem::CheckClient() 
+{
+  if (client_ == nullptr) 
+  {
+    return Status(Status::Code::INTERNAL, "blob client initialize failed.");
+  }
+  as::blob_client_wrapper bc(client_);
```
Comment (omnipo) [inline src/core/filesystem.cc]: is this not used?

File: src/core/filesystem.cc [hunk 2]
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
...
+    } else if (query.find("sig=") != std::string::npos) {
+      // Shared Access Signature
+      cred = std::make_shared<as::shared_access_signature_credential>(query);
+    } else {
+      cred = std::make_shared<as::anonymous_credential>();
+    }
+
+    account = std::make_shared<as::storage_account>(
+        account_name, cred, /* use_https */ true);
+    client_ = std::make_shared<as::blob_client>(account, 16);
```
Comment (omnipo) [inline src/core/filesystem.cc]: what is this '16'?

File: src/core/filesystem.cc [hunk 3]
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
...
+  std::string host_name, container, blob_path, query;
+  if (RE2::FullMatch(
+          path, as_regex_, &host_name, &container, &blob_path, &query)) {
+    size_t pos = host_name.rfind(".blob.core.windows.net");
+    std::string account_name;
+    if (pos != std::string::npos) {
+      account_name = host_name.substr(0, pos);
+    } else {
+      account_name = host_name;
+    }
```
Comment (omnipo) [inline src/core/filesystem.cc:766]: Does this convention have a doc?

File: src/core/filesystem.cc [hunk 4]
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
...
+
+ASFileSystem::ASFileSystem(const std::string& path)
+    : as_regex_("as://([^/]+)/([^/]+)/([^?]+)(\\?.*)?")
+{
+  const char* account_key = std::getenv("AZURE_STORAGE_KEY");
+  std::shared_ptr<as::storage_account> account = nullptr;
+  std::string host_name, container, blob_path, query;
+  if (RE2::FullMatch(
+          path, as_regex_, &host_name, &container, &blob_path, &query)) {
+    size_t pos = host_name.rfind(".blob.core.windows.net");
```
Comment (omnipo) [inline src/core/filesystem.cc:754]: how is rfind better than find?

File: src/core/filesystem.cc [hunk 5]
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
...
+{
+
+  as::blob_client_wrapper bc(client_);
+  std::string container, object_path;
+  RETURN_IF_ERROR(ParsePath(path, &container, &object_path));
+  using namespace azure::storage_lite;
+  std::ostringstream out_stream;
+  bc.download_blob_to_stream(container, object_path, 0, 0, out_stream);
+  if (errno!=0)
+  {
```
Comment (omnipo) [inline src/core/filesystem.cc]: let's follow the coding style here and move this brace to the line above

File: src/core/filesystem.cc [hunk 6]
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
...
+
+  std::string container, object;
+  RETURN_IF_ERROR(ParsePath(path, &container, &object));
+  as::blob_client_wrapper bc(client_);
+  auto blobs = bc.list_blobs_segmented(container, "/", "", object, 1);
+  if (errno != 0) {
+    return Status(Status::Code::INTERNAL, "Failed check file: " + path);
+  }
+  if (blobs.blobs.size() > 0) {
+        *exists = true;
```
Comment (omnipo) [inline src/core/filesystem.cc]: nit: indent

File: src/core/filesystem.cc [hunk 7]
```diff
@@ -1163,6 +1517,21 @@ GetFileSystem(const std::string& path, FileSystem** file_system)
 #endif  // TRITON_ENABLE_S3
   }
 
+  // Check if this is an Azure Storage path 
+  if (!path.empty() && !path.rfind("as://", 0)) {
+#ifndef TRITON_ENABLE_S3
```
Comment (omnipo) [inline src/core/filesystem.cc]: Should be ENABLE_AZURE?

File: src/core/filesystem.cc [hunk 8]
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
...
+      auto ret = DownloadFolder(container, blob_path, local_path);
+      if (!ret.IsOk()) 
+      {
+        return ret;
+      }
+    } else {
+      time_t last_modified;
+      bc.download_blob_to_file(container, blob_path, local_path, last_modified);
+      if (errno != 0) 
+      {
```
Comment (omnipo) [inline src/core/filesystem.cc]: brace :)

File: src/core/filesystem.cc (hunk 1 above)
Comment (spacedragon) [reply src/core/filesystem.cc]: I will remove it.

File: src/core/filesystem.cc (hunk 2 above)
Comment (spacedragon) [reply src/core/filesystem.cc]: max_concurrency

adding comments.

File: src/core/filesystem.cc (hunk 3 above)
Comment (spacedragon) [reply src/core/filesystem.cc:766]: No. `{account}.blob.core.windows.net` is the default assigned DNS name. But user can assign any domain to its storage address.

Adding a new env `AZURE_STORAGE_ACCOUNT` to cover this situation.

**Pull request closed without merging**